### DISCUSSION TITLE
IO Position Control

### DISCUSSION THREAD

**Discussion by heaton56** [The-OpenROAD-Project/OpenROAD-flow-scripts#634] [Q&A] [answered]
How are the IO positions controlled? 

I was testing progressively increasing the density of the gcd example design until it broke. 
When I reduced the CORE_AREA = 10.07 11.2 52.25 39.2, I noticed it no longer routed IO at the top. 

Hence the question: how to control the IO positions?

![image](https://user-images.githubusercontent.com/36714719/200102200-9244ad08-90ce-4093-81f4-f1a1199af997.png)

**Comment by maliberty** [comment]
You can read about set_io_pin_constraint in https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/ppl

**Reply by rovinski** [reply]
To expand, you can use put PPL commands into a script, then use that script in the flow by setting in your design config:
```
IO_CONSTRAINTS = path/to/your/script.tcl
```
You can also provide additional arguments to the pin placer with:
```
PLACE_PINS_ARGS = argument1 argument2 ...
```
There is also an option to use a pre-floorplanned DEF file by setting:
```
FLOORPLAN_DEF = /path/to/your/floorplan.def
```

**Comment by maliberty** [comment, marked as answer]
It places the IO to minimize wire length.  Since you have no instances near the top you it is unsurprising that you have no IOs there.
